### PR TITLE
[ONNX] Update TorchTensor implementation to handle fake mode

### PR DESCRIPTION
**Pull request by justinchuby**
Update TorchTensor implementation to handle fake mode better. Specifically, we disable fake mode before calling detach() etc. when getting the weights if it is already a real tensor so we do not lose it.

### CODE REVIEW

File: torch/onnx/_internal/exporter/_core.py [hunk 1]
```diff
@@ -126,13 +126,17 @@ def tobytes(self) -> bytes:
         # it avoids copying to a NumPy array
         import torch._subclasses.fake_tensor
 
-        if isinstance(self.raw, torch._subclasses.fake_tensor.FakeTensor):
+        with torch._subclasses.fake_tensor.unset_fake_temporarily():
+            # Disable any fake mode so calling detach() etc. will return a real tensor
```
Comment (fatcat-z) [inline torch/onnx/_internal/exporter/_core.py:130]: After doing this, does it mean the line 133 will be False forever?

Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:130]: Unless the original is fake already

Comment (fatcat-z) [reply torch/onnx/_internal/exporter/_core.py:130]: Could you please add a test in test_api.py for this?

**Review by fatcat-z — APPROVED** [review]
LGTM, thanks!

File: torch/onnx/_internal/exporter/_core.py (hunk 1 above)
Comment (justinchuby) [reply torch/onnx/_internal/exporter/_core.py:130]: Added tests

File: test/onnx/exporter/test_api.py [hunk 2]
```diff
@@ -294,5 +295,88 @@ def onnx_add(self: FLOAT, other: FLOAT) -> FLOAT:
...
+        real_model = Model()
+
+        with torch.onnx.enable_fake_mode():
+            onnx_program = torch.onnx.export(
+                real_model, (torch.tensor(1.0),), dynamo=True
+            )
+
+            assert onnx_program is not None
+            # Convert to model proto and back to trigger to_bytes method which serializes the tensor
+            # Note that even though we are calling .model_proto (equivalently .save()) in fake mode,
```
Comment (titaiwangms) [inline test/onnx/exporter/test_api.py:370]: I am confused by "we are calling .model_proto (equivalently .save()) in fake mode, ..." Isn't the only way for users to use fakemode is to .model_proto/.save() inside `torch.onnx.enable_fake_mode`, or there is another way?

Is this PR suggesting users to save and get model_proto under torch.onnx.enable_fake_mode()?

**Review by titaiwangms — APPROVED** [review]

File: test/onnx/exporter/test_api.py (hunk 2 above)
Comment (justinchuby) [reply test/onnx/exporter/test_api.py:370]: I don’t think so. It doesn’t matter where users call save, which is what this PR does. It used to be that users can only call save outside of the fake mode manager.

Comment (titaiwangms) [reply test/onnx/exporter/test_api.py:370]: I guess it's the first two test cases confuse me. I thought the first two cases mean that we should .model_proto/.save() inside the fake mode context manager, or it raises error?

Comment (justinchuby) [reply test/onnx/exporter/test_api.py:370]: I see. I can duplicate the model_proto calls to inside the context manager as well. It doesn't matter where it is called.

Comment (justinchuby) [reply test/onnx/exporter/test_api.py:379]: Please take another look. Thanks

**Review by titaiwangms — APPROVED** [review]
